src/Compilers/CSharp/Portable/Symbols/Synthesized/GeneratedNames.cs

```diff
@@ -120,30 +120,41 @@ internal static string MakeBaseMethodWrapperName(int uniqueId)
             return "<>n__" + StringExtensions.GetNumeral(uniqueId);
         }
 
-        internal static string MakeLambdaMethodName(string methodName, int methodOrdinal, int generation, int lambdaOrdinal)
+        internal static string MakeLambdaMethodName(string methodName, int methodOrdinal, int methodGeneration, int lambdaOrdinal, int lambdaGeneration)
         {
-            Debug.Assert(generation >= 0);
             Debug.Assert(methodOrdinal >= -1);
+            Debug.Assert(methodGeneration >= 0);
             Debug.Assert(lambdaOrdinal >= 0);
+            Debug.Assert(lambdaGeneration >= 0);
 
             // The EE displays the containing method name and unique id in the stack trace,
             // and uses it to find the original binding context.
-            return MakeMethodScopedSynthesizedName(GeneratedNameKind.LambdaMethod, methodOrdinal, generation, methodName, uniqueId: lambdaOrdinal);
+            return MakeMethodScopedSynthesizedName(GeneratedNameKind.LambdaMethod, methodOrdinal, methodGeneration, methodName, entityOrdinal: lambdaOrdinal, entityGeneration: lambdaGeneration);
         }
 
-        internal static string MakeLambdaCacheFieldName(int methodOrdinal, int generation, int lambdaOrdinal)
+        internal static string MakeLambdaCacheFieldName(int methodOrdinal, int generation, int lambdaOrdinal, int lambdaGeneration)
         {
             Debug.Assert(methodOrdinal >= -1);
             Debug.Assert(lambdaOrdinal >= 0);
 
-            return MakeMethodScopedSynthesizedName(GeneratedNameKind.LambdaCacheField, methodOrdinal, generation, uniqueId: lambdaOrdinal);
+            return MakeMethodScopedSynthesizedName(GeneratedNameKind.LambdaCacheField, methodOrdinal, generation, entityOrdinal: lambdaOrdinal, entityGeneration: lambdaGeneration);
         }
 
-        private static string MakeMethodScopedSynthesizedName(GeneratedNameKind kind, int methodOrdinal, int generation, string methodNameOpt = null, string suffix = null, int uniqueId = -1, bool isTypeName = false)
+        private static string MakeMethodScopedSynthesizedName(
+            GeneratedNameKind kind,
+            int methodOrdinal, 
+            int methodGeneration, 
+            string methodNameOpt = null,
+            string suffix = null, 
+            int entityOrdinal = -1, 
+            int entityGeneration = -1,
+            bool isTypeName = false)
         {
             Debug.Assert(methodOrdinal >= -1);
-            Debug.Assert(generation >= 0 || generation == -1 && methodOrdinal == -1);
-            Debug.Assert(uniqueId >= -1);
+            Debug.Assert(methodGeneration >= 0 || methodGeneration == -1 && methodOrdinal == -1);
+            Debug.Assert(entityOrdinal >= -1);
```

Comment:
Could this be strengthened to Debug.Assert((methodGeneration >= 0) == (methodOrdinal >= 0))?